base_layer/core/src/base_node/sync/header_sync/synchronizer.rs

```diff
@@ -245,59 +249,69 @@ impl<'a, B: BlockchainBackend + 'static> HeaderSynchronizer<'a, B> {
             .await?;
-        match sync_status {
-            SyncStatus::InSync | SyncStatus::WereAhead => {
-                let metadata = self.db.get_chain_metadata().await?;
-                if metadata.height_of_longest_chain() < header_tip_height {
+        let header_sync_status = self
+            .determine_sync_status(&sync_peer.to_string(), best_header, best_block, peer_response.clone())
+            .await?;
+
+        match header_sync_status.clone() {
+            HeaderSyncStatus::InSyncOrAhead => {
+                if best_block_height < best_header_height {
                     debug!(
                         target: LOG_TARGET,
                         "Headers are in sync at height {} but tip is {}. Proceeding to archival/pruned block sync",
-                        header_tip_height,
-                        metadata.height_of_longest_chain()
+                        best_header_height,
+                        best_block_height
                     );
-                    Ok(())
+
+                    Ok(AttemptSyncResult {
+                        headers_returned: peer_response.headers.len() as u64,
+                        fork_hash_index: peer_response.fork_hash_index,
+                        header_sync_status,
+                    })
                 } else {
-                    // Check if the metadata that we had when we decided to enter header sync is behind the peer's
-                    // claimed one. If so, our chain has updated in the meantime and the sync peer
-                    // is behaving.
-                    if self.local_metadata.accumulated_difficulty() <=
-                        sync_peer.claimed_chain_metadata().accumulated_difficulty()
-                    {
-                        debug!(
-                            target: LOG_TARGET,
-                            "Local blockchain received a better block through propagation at height {} (was: {}). \
-                             Proceeding to archival/pruned block sync",
-                            metadata.height_of_longest_chain(),
-                            self.local_metadata.height_of_longest_chain()
-                        );
-                        return Ok(());
-                    }
-                    debug!(
+                    // We will only attempt sync if the our accumulated difficulty is less than the peer's claimed
+                    // accumulated difficulty, thus this is adverse behaviour form the peer.
+                    warn!(
                         target: LOG_TARGET,
                         "Headers and block state are already in-sync (Header Tip: {}, Block tip: {}, Peer's height: \
-                         {})",
-                        header_tip_height,
-                        metadata.height_of_longest_chain(),
+                         {}), peer has lied about chain metadata or did not want to provide headers",
```

Comment:
This is not true. We only checked if our best_block < our best_headers

Reply:
Ok, I provided better data for a better decision

Reply:
Please provide a link to the new code or more detail about what you did, I can't find it among the other changes

Reply:
The conditions are now tested separately and both will be ban-able offences.
See
```rust
        // Basic sanity check about the peer's claimed chain metadata
        if chain_split_result.peer_accumulated_difficulty < sync_peer.claimed_chain_metadata().accumulated_difficulty()
```
and
```rust
        match header_sync_status.clone() {
            HeaderSyncStatus::InSyncOrAhead => {
                // Our POW is less than the peer's POW, as verified before the attempted header sync, therefore, if the
                // peer did not supply any headers and we know we are behind based on the peer's claimed metadata, then
                // we can ban the peer.
                if best_header.height() == best_block.height() {
```